shared-module/audiofilters/Distortion.c

```diff
@@ -0,0 +1,374 @@
+    bool single_channel_output,
+    uint8_t channel) {
+
+    memset(self->buffer[0], 0, self->buffer_len);
+    memset(self->buffer[1], 0, self->buffer_len);
+}
+
+bool common_hal_audiofilters_distortion_get_playing(audiofilters_distortion_obj_t *self) {
+    return self->sample != NULL;
+}
+
+void common_hal_audiofilters_distortion_play(audiofilters_distortion_obj_t *self, mp_obj_t sample, bool loop) {
+    // When a sample is to be played we must ensure the samples values matches what we expect
+    // Then we reset the sample and get the first buffer to play
+    // The get_buffer function will actually process that data
+
+    if (audiosample_sample_rate(sample) != self->sample_rate) {
+        mp_raise_ValueError_varg(MP_ERROR_TEXT("The sample's %q does not match"), MP_QSTR_sample_rate);
+    }
+    if (audiosample_channel_count(sample) != self->channel_count) {
+        mp_raise_ValueError_varg(MP_ERROR_TEXT("The sample's %q does not match"), MP_QSTR_channel_count);
+    }
+    if (audiosample_bits_per_sample(sample) != self->bits_per_sample) {
+        mp_raise_ValueError_varg(MP_ERROR_TEXT("The sample's %q does not match"), MP_QSTR_bits_per_sample);
+    }
+    bool single_buffer;
+    bool samples_signed;
+    uint32_t max_buffer_length;
+    uint8_t spacing;
+    audiosample_get_buffer_structure(sample, false, &single_buffer, &samples_signed, &max_buffer_length, &spacing);
+    if (samples_signed != self->samples_signed) {
+        mp_raise_ValueError_varg(MP_ERROR_TEXT("The sample's %q does not match"), MP_QSTR_signedness);
+    }
+
+    self->sample = sample;
+    self->loop = loop;
+
+    audiosample_reset_buffer(self->sample, false, 0);
+    audioio_get_buffer_result_t result = audiosample_get_buffer(self->sample, false, 0, (uint8_t **)&self->sample_remaining_buffer, &self->sample_buffer_length);
+
+    // Track remaining sample length in terms of bytes per sample
+    self->sample_buffer_length /= (self->bits_per_sample / 8);
+    // Store if we have more data in the sample to retrieve
+    self->more_data = result == GET_BUFFER_MORE_DATA;
+
+    return;
+}
+
+void common_hal_audiofilters_distortion_stop(audiofilters_distortion_obj_t *self) {
+    // When the sample is set to stop playing do any cleanup here
+    self->sample = NULL;
+    return;
+}
+
+static mp_float_t db_to_linear(mp_float_t value) {
+    return expf(value * MICROPY_FLOAT_CONST(0.11512925464970228420089957273422));
+}
+
+audioio_get_buffer_result_t audiofilters_distortion_get_buffer(audiofilters_distortion_obj_t *self, bool single_channel_output, uint8_t channel,
+    uint8_t **buffer, uint32_t *buffer_length) {
+
+    if (!single_channel_output) {
+        channel = 0;
+    }
+
+    // get the effect values we need from the BlockInput. These may change at run time so you need to do bounds checking if required
+    mp_float_t drive = MIN(MAX(synthio_block_slot_get(&self->drive), 0.0), 1.0);
```

Comment:
as detailed in https://github.com/adafruit/circuitpython/issues/9872 the distortion class needs to take responsibility for ticking the blocks it uses.

Reply:
I'd like to wait until we have a clear resolution for block ticking outside of synthio before including it within this PR. I will keep this conversation open in case that occurs before this PR is merge-able.